overlord/devicestate/devicemgr.go

```diff
@@ -2214,3 +2214,123 @@ func (m *DeviceManager) RemoveRecoveryKeys() error {
+	// requirements from the model vs the capabilities of the device
+	PolicyErr error
+}
+
+var secbootCheckTPMKeySealingSupported = secboot.CheckTPMKeySealingSupported
+
+// checkEncryption verifies whether encryption should be used based on the
+// model grade and the availability of a TPM device or a fde-setup hook
+// in the kernel.
+func (m *DeviceManager) checkEncryption(st *state.State, deviceCtx snapstate.DeviceContext) (secboot.EncryptionType, error) {
+	model := deviceCtx.Model()
+
+	// TODO: add error checking here and return proper error once
+	// all testsy mock KernelInfo
+	kernelInfo, _ := snapstate.KernelInfo(st, deviceCtx)
+	var gadgetInfo *gadget.Info
+	if gadgetSnapInfo, err := snapstate.GadgetInfo(st, deviceCtx); err == nil {
+		gadgetInfo, err = gadget.ReadInfo(gadgetSnapInfo.MountDir(), nil)
+		if err != nil {
+			return "", err
+		}
+	}
+
+	res, err := m.checkEncryptionAndRequirements(model, kernelInfo, gadgetInfo)
+	if err != nil {
+		return "", err
+	}
+	return res.Type, res.PolicyErr
+}
+
+func (m *DeviceManager) checkEncryptionAndRequirements(model *asserts.Model, kernelInfo *snap.Info, gadgetInfo *gadget.Info) (res EncryptionRequirements, err error) {
+	secured := model.Grade() == asserts.ModelSecured
+	dangerous := model.Grade() == asserts.ModelDangerous
+	encrypted := model.StorageSafety() == asserts.StorageSafetyEncrypted
+	preferUnencrypted := model.StorageSafety() == asserts.StorageSafetyPreferUnencrypted
+
+	res.Required = secured || encrypted
+
+	// check if we should disable encryption non-secured devices
+	// TODO:UC20: this is not the final mechanism to bypass encryption
+	if dangerous && osutil.FileExists(filepath.Join(boot.InitramfsUbuntuSeedDir, ".force-unencrypted")) {
+		res.Disabled = true
+	}
+
+	// check if encryption is available
+	var (
+		hasFDESetupHook    bool
+		checkEncryptionErr error = fmt.Errorf("internal error: no encryption check performed")
+	)
+	if kernelInfo != nil {
+		if hasFDESetupHook = hasFDESetupHookInKernel(kernelInfo); hasFDESetupHook {
+			res.Type, checkEncryptionErr = m.checkFDEFeatures()
+		}
+	}
+	// Note that having a fde-setup hook will disable the build-in
+	// secboot encryption
+	if !hasFDESetupHook {
+		checkEncryptionErr = secbootCheckTPMKeySealingSupported()
+		if checkEncryptionErr == nil {
+			res.Type = secboot.EncryptionTypeLUKS
+		}
+	}
+	res.Available = (checkEncryptionErr == nil)
+
+	// check if encryption is required
+	if checkEncryptionErr != nil {
+		if secured {
+			res.PolicyErr = fmt.Errorf("cannot encrypt device storage as mandated by model grade secured: %v", checkEncryptionErr)
+		} else if encrypted {
+			res.PolicyErr = fmt.Errorf("cannot encrypt device storage as mandated by encrypted storage-safety model option: %v", checkEncryptionErr)
+		}
+
+		if hasFDESetupHook {
+			logger.Noticef("not encrypting device storage as querying kernel fde-setup hook did not succeed: %v", checkEncryptionErr)
```

Comment:
the logging here should also be moved one level up, that's why I propose UnavailableWarning in the struct